mabwiser/treebandit.py

```diff
@@ -0,0 +1,163 @@
+            self.arm_to_tree[arm].fit(arm_contexts, arm_rewards)
+
+            # For each leaf, keep a list of rewards
+            # DecisionTreeClassifier's apply() method returns the indices of the nodes in the tree
+            # that the specified contexts lead to. Therefore, the indices returned are not necessarily
+            # consecutive/follow numerical order, but are always representative of leaf nodes.
+            leaf_indices = self.arm_to_tree[arm].apply(arm_contexts)
+
+            # Use set() to create a list of unique indices
+            # These indices represent the leaves reached via the given contexts
+            unique_leaf_indices = set(leaf_indices)
+
+            for index in unique_leaf_indices:
+                # Get rewards list for each leaf
+                self.arm_to_rewards[arm][index] = arm_rewards[leaf_indices == index]
+
+    def _predict_contexts(self, contexts: np.ndarray, is_predict: bool,
+                          seeds: Optional[np.ndarray] = None, start_index: Optional[int] = None) -> List:
+
+        # Get local copy of arm_to_tree, arm_to_expectation, arm_to_rewards, and arms
+        # to minimize communication overhead between arms (processes) using shared objects
+        arm_to_tree = deepcopy(self.arm_to_tree)
+        arm_to_rewards = deepcopy(self.arm_to_rewards)
+        arm_to_expectation = deepcopy(self.arm_to_expectation)
+        arms = deepcopy(self.arms)
+
+        # Create an empty list of predictions
+        predictions = [None] * len(contexts)
+        for index, row in enumerate(contexts):
+            # Each row needs a separately seeded rng for reproducibility in parallel
+            rng = np.random.RandomState(seed=seeds[index])
+
+            for arm in arms:
+
+                # Copy the row rng to the deep copied model in arm_to_tree
+                arm_to_tree[arm].rng = rng
+
+                # If there was prior data for this arm, get expectation for arm
+                if arm_to_rewards[arm]:
+                    # Get leaf index for that context
+                    leaf_index = arm_to_tree[arm].apply([row])[0]
+
+                    # Get the rewards list of that leaf
+                    leaf_rewards = arm_to_rewards[arm][leaf_index]
+
+                    # Create leaf lp
+                    leaf_lp = self._get_leaf_lp(arm)
+
+                    # Leaf LP: fit the same arm decision with the leaf rewards
+                    leaf_lp.fit(np.asarray([arm] * len(leaf_rewards)), leaf_rewards)
+
+                    # Leaf LP: predict expectation
+                    arm_to_expectation[arm] = leaf_lp.predict_expectations()[arm]
+
+            if is_predict:
+                predictions[index] = argmax(arm_to_expectation)
+            else:
+                predictions[index] = arm_to_expectation.copy()
+
+        # Return list of predictions
+        return predictions
+
+    def _uptake_new_arm(self, arm: Arm, binarizer: Callable = None, scaler: Callable = None):
```

Comment:
do you have a corresponding unit test, where you start a TreeBandit and then add a new arm to it? 

Preferably, if you have a verified behavior in a multi-arm scenario, you can drop one arm, and build that scenario again starting with less arms and then adding that arm. The results should match?

Reply:
test_add_arm tests adding the new arm, but I can add more test cases to check matching the results.

Reply:
I've realized a problem with this. If we want to add a new arm, we can add it, but we cannot partial_fit new context for that arm since TreeBandit doesn't do traditional partial fitting. So actually the issue here is that our partial_fit function only updates the leaves' list rewards, but it can't update for an arm that hasn't received any context/doesn't have a fitted tree yet.

Reply:
Wow, that's subtle --great catch! 

In that case, can we plz raise a NotImplementedError in this _uptake_new_arm(). Add a test to verify that this operation is not supported, yet. 

Add your explanation/comment on why this is not implemented in this function.

Reply:
I added some code instead that tries to partial fit and if it realizes that the tree has not yet been fitted, it fits that specific tree with the given context. Hope that's fine!

Reply:
I like that approach, it's what I would expect as an end-user to happen

Reply:
Let me comment on this in the partial_fit() side